### PR TITLE
multi-batch normalization

### PR DESCRIPTION
**Pull request by klevdiamanti**
# Normalization of multiple projects
**Problem:** When normalizing mutiple projects we had to manually perform the pair-wise normalizations.

**Solution:** This function allows us to provide a pre-defined nested data structure of data frames that will perform all normalizations (bridge and subset) and output a fully normalized NPX dataset.

**Key Features:**

* New multi-project normalization function olink_normalization_n
* No need for bridge samples to be named the same in both proejcts to be normalized
* New function that explicitly performs bridge normalization olink_normalization_bridge
* New function that explicitly performs subset normalization olink_normalization_subset

## Checklist
- [X] Make sure you are requesting to **pull a feature/bugfix branch** (right side). This should not be main or develop.
- [X] Make sure you are make a pull request against either **main or develop** (left side). (Requesting to main should be reserved for bugfixes and new releases)
- [X] Add or update unit tests (if applicable)
- [X] Check your code with any unit tests (Run devtools::check() locally)
- [X] Add neccessary documentation (if applicable)

## Type of changes

What type of changes does your code introduce?

- [ ] Bugfix (non-breaking change which fixes an issue) (link the issue on the right)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code style update (formatting, renaming)
- [X] Documentation Update 
- [ ] Other (explain)

## Further comments

- I have not had time to add tests yet, but will do so asap.
- This is a wrapper of the current function olink_normalization, which has to be rewritten an optimized. The main idea is that this function will be depreacated and replaced by olink_normalization_bridge, olink_normalization_subset and olink_normalization_median.
- This addresses #187
- Issue #186 will be addressed in the next update(s).

### CODE REVIEW

**Comment by klevdiamanti** [comment]
You can use OlinkAnalyze/inst/extdata/complex_dataset.rds or subsets of it to test the function.

File: OlinkAnalyze/R/olink_normalization_n.R [hunk 1]
```diff
@@ -0,0 +1,892 @@
...
+#'    supported.
+#'    \item{normalize_to:} a character array pointing to the project this
+#'    project is to be normalized to. Elements of the array should be
+#'    exclusively from the "order" column. Elements of the array may be
+#'    comma-separated if the project is to be normalized to multiple projects.
+#' }
+#'
+#' @param norm_schema A tibble with more than 1 rows and (strictly) the
+#' following columns: "order", "name", "data", "samples", "normalization_type",
+#' "normalize_to". See above for details of the structure of the data frame
```
Comment (Orbmac) [inline OlinkAnalyze/R/olink_normalization_n.R:66]: See below

"As the Argument section are above the details section when printed"

File: OlinkAnalyze/R/olink_normalization_n.R [hunk 2]
```diff
@@ -0,0 +1,892 @@
...
+#'    \item{2.} project 2 normalizes to project 1, and project 3 normalizes to
+#'    the combined data frame of projects 1 and 2 (that is already normalized).
+#' }
+#'
+#' The function can also handle a mixed schema of bridge and subset
+#' normalizations.
+#'
+#' Specifications of the normalization schema data frame:
+#' \itemize{
+#'    \item{order:} should strictly be a numeric or integer array with unique
```
Comment (Orbmac) [inline OlinkAnalyze/R/olink_normalization_n.R:29]: Looking at the examples (Havent reviewd code yet, dont this always have to be an integer? Remove the "a numeric or" part?

**Comment by Orbmac** [comment]
I might be missing something here but from how I understand the function (and how i understand the "normal" olink_normalization, shoudn't this two functions give the same resuts?


complex_small<-complex_dataset[1:4,]
olink_normalization_n(complex_small)->x

complex_small$data$DF1b->df1
complex_small$data$DF2b->df2

complex_small$samples$DF2b$DF1 ->df11_bridge
complex_small$samples$DF2b$DF2 ->df12_bridge

normal12bridge<-olink_normalization(df1, df2, 
                    overlapping_samples_df1 = df11_bridge,
                    overlapping_samples_df2 = df12_bridge,
                    df1_project_nr = '1',
                    df2_project_nr = '2',
                    reference_project = '1')

new12bridge<-x %>% filter(Project %in% c('DF1b','DF2b'))

**Comment by klevdiamanti** [comment]
> I might be missing something here but from how I understand the function (and how i understand the "normal" olink_normalization, shoudn't this two functions give the same resuts?
> 
> complex_small<-complex_dataset[1:4,] olink_normalization_n(complex_small)->x
> 
> complex_small$data$DF1b->df1 complex_small$data$DF2b->df2
> 
> complex_small$samples$DF2b$DF1 ->df11_bridge complex_small$samples$DF2b$DF2 ->df12_bridge
> 
> normal12bridge<-olink_normalization(df1, df2, overlapping_samples_df1 = df11_bridge, overlapping_samples_df2 = df12_bridge, df1_project_nr = '1', df2_project_nr = '2', reference_project = '1')
> 
> new12bridge<-x %>% filter(Project %in% c('DF1b','DF2b'))

I think it does work. In your example you have performed subset normalization when you directly used `olink_normalization` as you have provided both `overlapping_samples_df1` and `overlapping_samples_df2`. See example below (based on your code) for how we would do it:

```complex_small<-complex_dataset[1:4,]
olink_normalization_n(complex_small)->x

complex_small$data$DF1b->df1
complex_small$data$DF2b->df2

complex_small$samples$DF2b$DF1 ->df11_bridge
complex_small$samples$DF2b$DF2 ->df12_bridge

# temprorary df to allow changing SampleID
df_bridge_samples <- dplyr::tibble(DF1 = complex_small$samples$DF2b$DF1,
                                   DF2 = complex_small$samples$DF2b$DF2)

# replace SampleID of df1 with the SampleID of bridge samples from df2.
# this is so that the two df have the same SampleID for bridge samples
# as required by the olink_normalization function
df1 <- df1 |> 
  dplyr::left_join(df_bridge_samples, by = c('SampleID' = 'DF1')) |> 
  dplyr::mutate(SampleID = dplyr::if_else(is.na(DF2), SampleID, DF2)) |> 
  dplyr::select(-DF2)

normal12bridge<-olink_normalization(df1 = df1,
                                    df2 = df2,
                                    overlapping_samples_df1 = df12_bridge,
                                    df1_project_nr = '1',
                                    df2_project_nr = '2',
                                    reference_project = '1')

# change SampleID of df1 back to the original
normal12bridge <- normal12bridge |> 
  dplyr::left_join(df_bridge_samples, by = c('SampleID' = 'DF2')) |> 
  dplyr::mutate(SampleID = dplyr::if_else(!is.na(DF1) & Project == "1", DF1, SampleID)) |> 
  dplyr::select(-DF1)

new12bridge<-x %>% filter(Project %in% c('DF1b','DF2b'))

# this just to showcase that it works
combo12bridge <- normal12bridge |> 
  dplyr::left_join({
    new12bridge |> 
      dplyr::select(SampleID, PlateID, OlinkID, NPXnew = NPX, LODnew = LOD,
                    Projectnew = Project, Adj_factornew = Adj_factor)
  }, by = c('SampleID', 'PlateID', 'OlinkID')) |> 
  dplyr::select(SampleID:QC_Warning, Normalization:Assay_Warning,
                NPX, NPXnew, LOD, LODnew,
                Project, Projectnew, Adj_factor, Adj_factornew)
```

**Review by Orbmac — COMMENTED** [review]
Tried to play around with the function and from my testing at least looks good. A bit hassle to set up but I can see when one is more used to it will be very nice. Other than my comments not much else to add for now.

File: OlinkAnalyze/R/olink_normalization_n.R [hunk 3]
```diff
@@ -0,0 +1,892 @@
...
+        as.integer()
+    },
+    1:nrow(norm_schema))) {
+    return_message <- paste("order has to be a unique integer or numeric",
+                            "identifier of each row!")
+  }
+
+  # Check data
+  if (!is.list(norm_schema$data) ||
+      !all(sapply(norm_schema$data, is.data.frame))) {
```
[the remark this answers is not in the document]
Comment (klevdiamanti) [reply OlinkAnalyze/R/olink_normalization_n.R:849]: An is_npx function is in the works but it is going to take a while to be ready. In the meantime we have to have this functions ready.

**Review by Orbmac — APPROVED** [review]
Look over all the tests and looks to be doing what it say. Great work :)

**Pull request merged**